chore(lint): `oxlint --type-aware`

Boshen opened the pull request:
To enable type aware linting:

1. `pnpm add -w -D oxlint-tsgolint`
2. add `--type-aware` to cli
3. enable the rules

```
    "typescript/no-floating-promises": "error",
    "typescript/no-misused-promises": "error"
```

(or in cli `-D typescript/no-floating-promises -D typescript/no-misused-promises`)

<img width="1768" height="1704" alt="image" src="https://github.com/user-attachments/assets/8392a24f-8029-4f41-abd7-14d400de3cfb" />

hyf0 commented on packages/bench/src/types.d.ts:
```diff
@@ -16,15 +27,15 @@ export interface BenchSuite {
-    | RolldownOptions
-    | { name: string; options: RolldownOptions }[];
-  rollupOptions?: RollupOptions;
+    | RolldownOptionsWithSingleOutput
```
I don't know the context, but it seems to increase the complexity without any benefit. Could you explain why we need to do this?

hyf0 replied:
Or just disable the rule and left a todo for future investigation.

shulaoda replied:
That’s because the `output` field in `RolldownOptions` can be either `OutputOptions` or `OutputOptions[]`, while our `BenchSuite` assumes it’s just a single `OutputOptions`.
https://github.com/rolldown/rolldown/blob/d60a0e9fedd33d259da1e9201ece87fa15bd3c4c/packages/bench/src/run-bundler.ts#L51-L72

hyf0 commented on packages/rolldown/src/api/dev/index.ts:
```diff
@@ -1,3 +1,4 @@
 import { DevEngine } from './dev-engine';
 
-export var dev: typeof DevEngine.create = DevEngine.create;
+export const dev: typeof DevEngine.create = (...args) =>
```
And why we want to do this?

shulaoda replied:
The typing issue with `this`.

hyf0 commented on packages/rolldown/src/api/watch/watch-emitter.ts:
```diff
@@ -136,8 +136,9 @@ export class WatcherEmitter {
-        default:
-          throw new Error(`Unknown event: ${event}`);
+        default: {
+          throw new Error(`Unknown event: ${event.eventKind()}`);
```
Redundant and lacks informtion.

hyf0 commented on packages/rolldown/src/cli/commands/bundle.ts:
```diff
@@ -25,7 +25,7 @@ export async function bundleWithConfig(
 
   if (!config) {
-    logger.error(`No configuration found at ${config}`);
+    logger.error(`No configuration found at ${configPath}`);
```
This one actually look right! Good catch! 👍

hyf0 commented on packages/rolldown/src/plugin/bindingify-hook-filter.ts:
```diff
@@ -189,7 +189,7 @@ function bindingifyFilterExprImpl(
     }
     default:
-      throw new Error(`Unknown filter expression: ${expr}`);
+      throw new Error(`Unknown filter expression: ${expr as string}`);
```
https://github.com/rolldown/rolldown/pull/5660/files#r2453829545

hyf0 commented on packages/rolldown/src/plugin/bindingify-plugin-hook-meta.ts:
```diff
@@ -21,7 +21,7 @@ function bindingPluginOrder(
       return undefined;
     default:
-      throw new Error(`Unknown plugin order: ${order}`);
+      throw new Error(`Unknown plugin order: ${order as string}`);
```
https://github.com/rolldown/rolldown/pull/5660/files#r2453829545

hyf0 commented on packages/rolldown/src/utils/bindingify-input-options.ts:
```diff
@@ -195,7 +195,8 @@ function bindingifyExperimental(
       default:
         throw new Error(
-          `Unexpected chunkModulesOrder: ${experimental.chunkModulesOrder}`,
+          `Unexpected chunkModulesOrder: ${experimental
```
https://github.com/rolldown/rolldown/pull/5660/files#r2453829545

hyf0 commented on packages/rolldown/src/utils/bindingify-output-options.ts:
```diff
@@ -127,7 +127,7 @@ function bindingifyFormat(
       return 'umd';
     }
     default:
```
https://github.com/rolldown/rolldown/pull/5660/files#r2453829545

hyf0 commented on packages/rolldown/src/utils/bindingify-output-options.ts:
```diff
@@ -145,7 +145,7 @@ function bindingifySourcemap(
       return 'hidden';
     default:
-      throw new Error(`unknown sourcemap: ${sourcemap}`);
+      throw new Error(`unknown sourcemap: ${sourcemap as string}`);
```
https://github.com/rolldown/rolldown/pull/5660/files#r2453829545

hyf0 commented on packages/rolldown/src/utils/normalize-string-or-regex.ts:
```diff
@@ -20,7 +20,7 @@ export function normalizedStringOrRegex<
 // For https://github.com/microsoft/TypeScript/issues/17002
 function isReadonlyArray<T extends unknown[] | readonly unknown[]>(
-  input: T | unknown,
+  input: unknown,
```
Could you check if this function still infers a correct type for return type?

shulaoda replied:
Yes, I think it is

hyf0 commented on packages/test-dev-server/src/dev-server.ts:
```diff
@@ -185,7 +185,7 @@ class DevServer {
           break;
         default:
-          throw new Error(`Unknown update type: ${update}`);
+          throw new Error(`Unknown update type: ${update as any}`);
```
https://github.com/rolldown/rolldown/pull/5660/files#r2453829545

shulaoda commented:
### Merge activity

* **Oct 24, 6:45 AM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Oct 24, 6:55 AM UTC**: `shulaoda` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Oct 24, 7:10 AM UTC**: Merged by the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).

Boshen committed:
chore(lint): `oxlint --type-aware` (#5660)

To enable type aware linting:

1. `pnpm add -w -D oxlint-tsgolint`
2. add `--type-aware` to cli
3. enable the rules

```
    "typescript/no-floating-promises": "error",
    "typescript/no-misused-promises": "error"
```

(or in cli `-D typescript/no-floating-promises -D typescript/no-misused-promises`)

<img width="1768" height="1704" alt="image" src="https://github.com/user-attachments/assets/8392a24f-8029-4f41-abd7-14d400de3cfb" />

hyf0 commented on .github/workflows/ci.yml:
```diff
@@ -173,6 +173,12 @@ jobs:
+      - name: Build @rolldown/pluginutils
+        run: pnpm --filter '@rolldown/pluginutils' build
+
+      - name: Build rolldown
```
Got a problem here.

I'm not sure the cause, but it seems we're running latest rolldown's ts code with loading old binary, which causes a problem when mofiying binary code.

The key here is that rolldown builds itself with rolldown's source code.
